### PR TITLE
#1134641313510243 Video quality option when sharing video projects

### PR DESCRIPTION
**Pull request by EkaterinaMaksimlyuk**

### CODE REVIEW

**Comment by EkaterinaMaksimlyuk** [comment]
@EugeneButusov  I removed the changes for YouTube,   as the documentation warns "Calling this function does not guarantee that the playback quality will actually change."

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js [hunk 1]
```diff
@@ -192,26 +229,67 @@
...
-                loadHlsJs(media, function(hls) {
+                loadHlsJs(media, function (hls) {
                   if(Hls.isSupported()) {
                     hls.on(Hls.Events.ERROR, function (error, data) {
                       // fallback to default media source
                       if (data.type === 'networkError') {
                         media.src = fallbackMedia;
                       }
                     });
+                    hls.on(Hls.Events.MEDIA_ATTACHED, function () {
```
Comment (aleksey-lav) [inline wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: is this for appple devices ?

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js [hunk 2]
```diff
@@ -192,26 +229,67 @@
...
+                          return q;
+                        });
+                        bitrates.push({ resolution: "auto", value: "auto" });
+                        media.qualities = bitrates;
+                      } else {
+                        media.qualities = [];
+                      }
+                      media.dispatchEvent("loadedbitrate");
+                      updateQuality = function (quality) {
+                        hls.currentLevel = quality === "auto" ? -1 : quality;
```
Comment (aleksey-lav) [inline wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: why `-1` ?

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js (hunk 1 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: yes

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js (hunk 2 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: from the documentation

**Review by aleksey-lav — APPROVED** [review]

File: wrappers/common/popcorn._MediaElementProto.js [hunk 3]
```diff
@@ -181,64 +183,84 @@
...
       },
 
       defaultPlaybackRate: {
-        get: function() {
+        get: function () {
           return 1.0;
         },
         configurable: true
       },
+      qualities: {
```
Comment (EugeneButusov) [inline wrappers/common/popcorn._MediaElementProto.js:234]: `quality` and `qualities` are very similar naming and it's easy to do a mistake during the using of this code. Better to set first as `currentQuality` or `activeQuality` or `selectedQuality`.

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js [hunk 4]
```diff
@@ -110,30 +114,63 @@
...
+          impl.quality = val;
+          if (updateQuality) {
+            updateQuality(impl.quality);
+          }
+        },
+        configurable: true
+      },
       src: {
-        get: function() {
+        get: function () {
```
Comment (EugeneButusov) [inline wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: these spaces are out of our codestyle approach I think.

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js [hunk 5]
```diff
@@ -192,26 +229,67 @@
...
+                      });
+                      bitrates.push({ resolution: "auto", value: "auto" });
+                      media.qualities = bitrates;
+                    } else {
+                      media.qualities = [];
+                    }
+                    media.quality = player.getQualityFor('video');
+                    media.dispatchEvent("loadedbitrate");
+                    updateQuality = function (quality) {
+                      if (quality === "auto") {
```
Comment (EugeneButusov) [inline wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: string literal is a way to make a typo, may be make constant for this literal, or use something like -1 as literal when auto. in addition to that http://cdn.dashjs.org/latest/jsdoc/module-MediaPlayer.html#setQualityFor, argument of setQualityFor is number, and you can have in quality string (auto) or number. Mixing type of variable is not seeming too reliable.

Comment (EugeneButusov) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: and the similar I see for HLS below.

**Review by EugeneButusov — CHANGES_REQUESTED** [review]

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js [hunk 6]
```diff
@@ -192,26 +229,67 @@
...
+                        q.value = idx;
+                        return q;
+                      });
+                      bitrates.push({ resolution: "auto", value: "auto" });
+                      media.qualities = bitrates;
+                    } else {
+                      media.qualities = [];
+                    }
+                    media.quality = player.getQualityFor('video');
+                    media.dispatchEvent("loadedbitrate");
```
Comment (EugeneButusov) [inline wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: `bitrateloaded`.

**Review by EugeneButusov — CHANGES_REQUESTED** [review]

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js (hunk 6 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: done

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js (hunk 5 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: done

File: wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js (hunk 4 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/adaptive/popcorn.HTMLAdaptiveMediaElement.js]: done

File: wrappers/common/popcorn._MediaElementProto.js (hunk 3 above)
Comment (EkaterinaMaksimlyuk) [reply wrappers/common/popcorn._MediaElementProto.js:234]: done

**Review by EugeneButusov — APPROVED** [review]

**Pull request merged**